sql/presentation.go

```diff
@@ -0,0 +1,181 @@
+)
+
+// presentation constructs a presentationNode.
+func (*planner) presentation(n *parser.Select, p planNode, s *scanNode) (planNode, error) {
+	if n.Limit == nil && n.Distinct == nil {
+		return p, nil
+	}
+	// Mark sort node as not presenting the results.
+	sNode, ok := p.(*sortNode)
+	if ok {
+		sNode.dontPresent = true
+	}
+	var count, offset int64
+	if n.Limit != nil {
+		data := []struct {
+			name       string
+			src        parser.Expr
+			dst        *int64
+			defaultVal int64
+		}{
+			{"LIMIT", n.Limit.Count, &count, math.MaxInt64},
+			{"OFFSET", n.Limit.Offset, &offset, 0},
+		}
+
+		for _, datum := range data {
+			if datum.src == nil {
+				*datum.dst = datum.defaultVal
+			} else {
+				if parser.ContainsVars(datum.src) {
+					return nil, fmt.Errorf("argument of %s must not contain variables", datum.name)
+				}
+
+				normalized, err := parser.NormalizeExpr(datum.src)
+				if err != nil {
+					return nil, err
+				}
+				dstDatum, err := parser.EvalExpr(normalized)
+				if err != nil {
+					return nil, err
+				}
+
+				if dstDatum == parser.DNull {
+					*datum.dst = datum.defaultVal
+					continue
+				}
+
+				if dstDInt, ok := dstDatum.(parser.DInt); ok {
+					*datum.dst = int64(dstDInt)
+					continue
+				}
+
+				return nil, fmt.Errorf("argument of %s must be type %s, not type %s", datum.name, parser.DummyInt.Type(), dstDatum.Type())
+			}
+		}
+	}
+	pNode := &presentationNode{planNode: p, columns: p.Columns(), limit: (n.Limit != nil), count: count, offset: offset}
+
+	if n.Distinct != nil {
+		pNode.checksums = make(map[[md5.Size]byte]struct{})
+		pNode.distinctIdx = make([]int, 0, len(n.Distinct))
+		for _, expr := range n.Distinct {
+			index, err := s.maybeAddRender(expr)
+			if err != nil {
+				return nil, err
+			}
+			pNode.distinctIdx = append(pNode.distinctIdx, index)
+		}
+	}
+	return pNode, nil
+}
+
+type presentationNode struct {
+	planNode
+	// This map is surely going to overflow memory.
+	// TODO(vivek): Return an error once map has grown above a configured threshold.
+	// Don't use map when the results are sorted.
+	checksums map[[md5.Size]byte]struct{} // keep track of rows to eliminate duplicates
+	// lastChecksum   md5.Size
+	// Index into Distinct ON evaluated expressions.
+	distinctIdx    []int // expressions that identify duplicates.
+	columns        []string
+	limit          bool
+	count          int64
+	offset         int64
+	rowIndex       int64
+	outputRowIndex int64
+}
+
+func (n *presentationNode) Next() bool {
+	if n.limit && n.outputRowIndex >= n.count {
+		return false
+	}
+
+	for n.planNode.Next() {
+		if n.checksums != nil {
+			// Detect duplicates
+			var key string
+			if len(n.distinctIdx) == 0 {
+				key = n.Values().String()
+			} else {
+				v := n.planNode.Values()
+				for _, index := range n.distinctIdx {
+					d := v[index-1]
+					// Reduces the chance of finding false duplicates by adding ",".
+					key += d.String() + ","
+				}
+			}
+			ckSum := md5.Sum([]byte(key))
```

Comment:
Using md5 here scares me a little bit. And so does that comment about "reduces the chance of finding false duplicates". Rather than using md5, we could create a key using the `util/encoding` routines that is composed of the distinct on values.